packages/pkl.experimental.structuredRead/structuredRead.pkl

```diff
@@ -0,0 +1,276 @@
+        local maxPathLength = fails.map((f) -> f.path.join(pathSeparator).length).max
+        for (fail in fails) {
+          let (path = fail.path.join(pathSeparator))
+            "\t'\(path)'\(" ".repeat(maxPathLength - path.length)): \(fail.message)"
+        }
+      }.join("\n"))
+        throw(msg)
+
+local function longestPath(fails: List<Failure>) =
+  fails.map((f) -> f.path.join(pathSeparator).length).max
+
+local function findFailures(output: Mapping<String, Any?>|Dynamic): List<Failure> =
+  output.toMap().fold(List(), (acc: List<Failure>, _, value) ->
+    if (value is Failure)
+      acc.add(value)
+    else if (value is Mapping)
+      acc + findFailures(value)
+    else if (value is Dynamic)
+      acc + findFailures(value)
+    else
+      acc
+  )
+
+abstract class Failure {
+  message: String
+  path: List<String>
+}
+
+class ReadFailure extends Failure
+class CoerceFailure extends Failure
+
+local function ReadFail(path: List<String>, msg: String) = let (_path = path ) new ReadFailure {
+  message = msg
+  path = _path
+}
+
+local function CoerceFail(path: List<String>, msg: String) = let (_path = path ) new CoerceFailure {
+  message = msg
+  path = _path
+}
+
+local function parseArraySyntax(input: String): Listing = new {
+  when (input.startsWith("{") && input.endsWith("}")) {
+    ...input.drop(1).dropLast(1).split(",").map((e) -> e.trim())
+  } else {
+    input
+  }
+}
+
+local function applyToDynamic(path, default: Dynamic) =
+  if (default.toList().isEmpty && default.toMap().isEmpty)
+    // Do some special handling for empty dynamic types
+     let (potentialValue = readValue(path))
+       if (potentialValue is ReadFailure)
+         default
+       else if (potentialValue is String)
+         parseArraySyntax(potentialValue)
+       else
+         potentialValue ?? default
+  else
+    new Mapping {
+      for (name, defaultValue in default) {
+        when (defaultValue is Dynamic) {
+          [name] = applyToDynamic(path.add(name.toString()), defaultValue)
+        } else {
+          [name] = let (_path = path.add(name.toString()))
+            let (potentialValue = coerceInput(readValue(_path), defaultValue.getClass(), _path))
+            if (potentialValue is ReadFailure)
+              defaultValue
+            else
+              potentialValue
+        }
+      }
+    }.toMap().toDynamic()
+
+hidden classHandlers: Mapping<Class, (String) -> Any> = new {
+  [Int] = (value) ->
+    value.toIntOrNull()
+
+  [String] = (value) ->
+    value
+
+  [Float] = (value) ->
+    value.toFloatOrNull()
+
+  [Number] = (value) ->
+    value.toIntOrNull() ?? value.toFloatOrNull()
+
+  [Boolean] = (value) ->
+    value.toBooleanOrNull()
+
+  [Null] = (_) -> throw("""
+    Unable to perform structured read due to ambiguous type in schema.
+    This is normally caused by a schema property having default value of `null`, and no type signature.
+  """)
```

Comment:
Is the indent intended?
```suggestion
    Unable to perform structured read due to ambiguous type in schema.
    This is normally caused by a schema property having default value of `null`, and no type signature.
    """)
```

Reply:
Probably not. Bad habit from Go, I still assume that all this stuff will be auto-fixed on save